File: bluerov2_standalone/catkin_ws/src/sys_id/uuv_model.m
function [dx,y] = uuv_model(t,x,u,varargin)
% states
% x(1): x position 
% x(2): y position
% x(4): vehicle heading

% u(1) is heading-change
% u(2) is speed
% matrices = 
% A = [0.999683993461193,0.00107006380992925,-2.58895074646266;
%     3.93865064196743e-05,0.999865198478192,-0.290848899831410;
%     1.18474912533678e-06,-5.01931779057720e-06,0.583866645209362];
% B = [0.0369928875656454,0.844042316730837;
%     0.0163886082047795,0.446282095713212;
%     0.00728931565945887,0.00953122338961594];

A = [0.000110934757707122,	0.0176481545263511,	-0.361993031166798;
0.000113111902385930,	0.00162986090236893, 1.63321898403585;
2.82185649873063e-05,	-1.75925433470232e-05,	0.645980176182907];


B =[ 0.00588416436363635, 0.536394967123741;
-0.0240381022203886, 0.555059923037031;
-0.0112500920603916, 0.227799073448295];

y(1) = x(1);
y(2) = x(2);
y(3) = x(3);
xt = [x(1);x(2);x(3)];
ut = [u(1);u(2)];
dx = A* xt + B*ut;
end 

File: bluerov2_standalone/catkin_ws/src/rtreach/reach_src/uuv_model.c
#include "uuv_model.h"
#include "main.h"
#include "face_lift.h"
#include "util.h"
#include "simulate_uuv.h"
#include "uuv_safety.h"
#include <stdio.h>
#include <string.h> 
#include <stdlib.h>


// a note from the f1tenth simulator 
// the car is 0.5 m long in the x direction 
// 0.3 long in the y direction




// do face lifting with the given settings, iteratively improving the computation
// returns true if the reachable set of states is satisfactory according to the
// function you provide in LiftingSettings (reachedAtIntermediateTime, reachedAtFinalTime)

bool face_lifting_iterative_improvement_uuv(int startMs, LiftingSettings* settings, REAL heading, REAL speed);

// helper function to check safety
// bool check_safety(HyperRectangle* rect, double (*box)[2]); 


// function that stops simulation after two seconds
bool shouldStop(REAL state[NUM_DIMS], REAL simTime, void* p)
{
	bool rv = false;
    REAL maxTime = 2.0f;
    // stop if the maximum simulation time 
	if (simTime >= maxTime)
	{
		rv = true;

		REAL* stopTime = (REAL*)p;
		*stopTime = -1;
	}

	return rv;
}



// Simulation 
REAL getSimulatedSafeTime(REAL start[3],REAL heading,REAL speed)
{
	REAL stepSize = 0.02f;
	REAL rv = 0.0f;

	simulate_uuv(start, heading,speed,stepSize, shouldStop, (void*)&rv); // TODO: look here

	//DEBUG_PRINT("time until simulation reaches safe state = %f\n", rv);

	return rv;
}

// called on states reached during the computation
bool intermediateState(HyperRectangle* r)
{
	bool allowed = true;
	//const REAL FIFTEEN_DEGREES_IN_RADIANS = 0.2618;

	// bloat the box for the width of the car
	r->dims[0].min = r->dims[0].min  - 0.99;
	r->dims[0].max = r->dims[0].max  + 0.99;
	r->dims[1].min = r->dims[1].min  - 0.115;
	r->dims[1].max = r->dims[1].max  + 0.115;

	
	allowed = check_safety_obstacles(r);

	// reset it
	r->dims[0].min = r->dims[0].min  + 0.99;
	r->dims[0].max = r->dims[0].max  - 0.99;
	r->dims[1].min = r->dims[1].min  + 0.115;
	r->dims[1].max = r->dims[1].max  - 0.115;

	//if(!allowed)
	//	printf("unsafe....\n");
	return allowed;
}

// This function enumerates all of the corners of the current HyperRectangle and 
// returns whether or not any of the points lies outside of the ellipsoid
bool finalState(HyperRectangle* rect)
{

	return intermediateState(rect);
}




bool runReachability_uuv(REAL* start, REAL simTime, REAL wallTimeMs, REAL startMs,REAL heading, REAL speed)
{
	LiftingSettings set;
	for (int d = 0; d < NUM_DIMS; ++d)
	{
		set.init.dims[d].min = start[d];
		set.init.dims[d].max = start[d];
	}

	set.reachTime = simTime;
	set.maxRuntimeMilliseconds = wallTimeMs;

	REAL iss = set.reachTime;
	iss = iss * 0.10f;

	
	

	set.initialStepSize = iss; 
	set.maxRectWidthBeforeError = 100;

	set.reachedAtFinalTime = finalState;
	set.reachedAtIntermediateTime = intermediateState;
	set.restartedComputation = 0; //restartedComputation;

	return face_lifting_iterative_improvement_uuv(startMs, &set,heading, speed);
}



